File: elixir/test/one_time_password_direct_test.exs
# OneTimePassword direct test (offline, mock system.fetch)

defmodule Thesmsworks.OneTimePasswordDirectTest do
  use ExUnit.Case

  alias Voxgig.Struct, as: S
  alias Thesmsworks.Helpers, as: H

  # A recording live-mode SDK whose transport returns `canned`.
  defp mk(canned) do
    parent = self()

    fetch = fn url, _fetchdef ->
      send(parent, {:called, url})
      {S.jm(["status", 200, "statusText", "OK", "headers", S.jm([]), "json", fn -> canned end, "body", "mock"]), nil}
    end

    Thesmsworks.new(S.jm(["base", "http://localhost:8080", "system", S.jm(["fetch", fetch])]))
  end

  test "should direct load one_time_password" do
    canned = S.jm(["id", "direct01"])
    sdk = mk(canned)
    res = Thesmsworks.direct(sdk, H.deep(%{"path" => "/one_time_password/direct01", "method" => "GET", "params" => %{}}))
    assert S.getprop(res, "ok") == true
    assert H.to_int(S.getprop(res, "status")) == 200
    data = S.getprop(res, "data")
    assert S.ismap(data)
    assert S.getprop(data, "id") == "direct01"
    assert_received {:called, _url}
  end
end

File: cpp/test/one_time_password_direct_test.cpp
// Generated direct-call tests for the one_time_password entity (unit mode;
// a mock system.fetch records calls). Mirrors the rust/go TestDirect.

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct OneTimePasswordDirectSetup {
  std::shared_ptr<ThesmsworksSDK> client;
  Value calls;
  bool live = false;
};

static OneTimePasswordDirectSetup one_time_password_direct_setup(const Value& mockres) {
  Value calls = vlist();
  Value cshared = calls;

  vs::Injector mock_fetch = [cshared, mockres](vs::Injection&, const Value& args, const std::string&, const Value&) -> Value {
    Value url = vs::getelem(args, Value(int64_t(0)));
    Value init = vs::getelem(args, Value(int64_t(1)));
    cshared.as_list()->push_back(vmap({{"url", url}, {"init", init}}));
    Value data = is_nullish(mockres) ? vmap({{"id", Value("direct01")}}) : mockres;
    Value out = vmap();
    map_put(out, "status", Value(200));
    map_put(out, "statusText", Value("OK"));
    map_put(out, "headers", vmap());
    map_put(out, "json", json_thunk(data));
    return out;
  };

  Value opts = vmap({
    {"base", Value("http://localhost:8080")},
    {"system", vmap({{"fetch", Value(mock_fetch)}})}
  });
  auto client = std::make_shared<ThesmsworksSDK>(opts);

  OneTimePasswordDirectSetup s;
  s.client = client;
  s.calls = calls;
  s.live = false;
  return s;
}

static void one_time_password_direct_load() {
  auto setup = one_time_password_direct_setup(vmap({{"id", Value("direct01")}}));
  auto sk = is_control_skipped("direct", "direct-load-one_time_password", "unit");
  if (sk.first) { std::cerr << "skip\n"; return; }
  auto client = setup.client;

  Value params = vmap();
  setp(params, "messageid", Value("direct01"));

  Value result = client->direct(vmap({
    {"path", Value("otp/{messageid}")},
    {"method", Value("GET")},
    {"params", params}
  }));

  ASSERT_EQ_VAL(getp(result, "ok"), Value(true), "expected ok true");
  ASSERT_EQ(Helpers::toInt(getp(result, "status")), 200, "expected status 200");
  ASSERT_TRUE(!getp(result, "data").is_undef(), "expected data to be non-nil");
  {
    Value data = getp(result, "data");
    if (data.is_map()) {
      ASSERT_EQ_VAL(getp(data, "id"), Value("direct01"), "expected data.id to be direct01");
    }
    ASSERT_EQ((int)setup.calls.as_list()->size(), 1, "expected 1 call");
    Value call = (*setup.calls.as_list())[0];
    ASSERT_EQ_VAL(getp(getp(call, "init"), "method"), Value("GET"), "expected method GET");
    std::string url = as_str(getp(call, "url"));
    ASSERT_TRUE(url.find("direct01") != std::string::npos, "expected url to contain direct01");
  }
}

int main() {
  T_RUN(one_time_password_direct_load);
  return sdktest::summary("one_time_password_direct_test");
}
